### PR TITLE
8340565: Create separate index page for terms defined by the index tag

### PR DESCRIPTION
**Pull request by hns**
This change adds a conditional `search-tags.html` page to list all search tags defined by the `{@index ... }` tag. Since search tags work very similar to the `{@systemProperty ...}` tag and are internally represented by the same class I changed `SystemPropertiesWriter.java` into an abstract base class called `IndexItemListWriter.java`, with concrete subclasses for system properties and search tags.

JavaDoc also generates index items for documentation headings, these are not included in `search-tags.html`.

[This is the Search Tags page for JDK API docs](https://cr.openjdk.org/~hannesw/8340565/api.00/search-tags.html) (top-level files only).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340565](https://bugs.openjdk.org/browse/JDK-8340565): Create separate index page for terms defined by the index tag (**Enhancement** - P4)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21429/head:pull/21429` \
`$ git checkout pull/21429`

Update a local copy of the PR: \
`$ git checkout pull/21429` \
`$ git pull https://git.openjdk.org/jdk.git pull/21429/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21429`

View PR using the GUI difftool: \
`$ git pr show -t 21429`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21429.diff">https://git.openjdk.org/jdk/pull/21429.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21429#issuecomment-2403005606)
</details>

### CODE REVIEW

**Comment by jonathan-gibbons** [comment]
> JavaDoc also generates index items for documentation headings, these are not included in search-tags.html.

This seems worth discussing.   "Module Resolution" is one heading that could be indexed, while there are others on the same page that arguably should not be.
https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/module/package-summary.html

**Comment by hns** [comment]
Out of curiosity, I hacked up two writers that list search items derived from [`<h1>` tags][h1] and [`<h2>` tags][h2]. While  think these items are very valuable in search, the signal/noise ratio in listed form appears rather low even for those higher-level headings.

[h1]: https://cr.openjdk.org/~hannesw/8340565/api.hdr/h1-search-tags.html
[h2]: https://cr.openjdk.org/~hannesw/8340565/api.hdr/h2-search-tags.html

**Comment by jonathan-gibbons** [comment]
Just asking: is it publicly documented anywhere that headings go into the interactive search? 

I guess there is a reasonably simple self-consistent rule that explicit use of `{@index}` tags go into this new page, but other items (like headings) that go into interactive search do not.  I'm trying to stay clear of even hinting that headings are treated as-if they had an implicit `{@index}` tag.

Also, note that "top-level" headings in a traditional comment for a member declaration start at `<h3>`. (For Markdown comments, all top-level headings for all declarations start at level 1.)

**Comment by hns** [comment]
> Just asking: is it publicly documented anywhere that headings go into the interactive search?

Good question, I don't think so. I think it could/should be documented in the spec for the `{@index}` tag, maybe also the Standard Doclet output spec?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Navigation.java [hunk 1]
```diff
@@ -87,8 +87,9 @@ public enum PageMode {
         PACKAGE,
         PREVIEW,
         RESTRICTED,
-        SERIALIZED_FORM,
         SEARCH,
+        SEARCH_TAGS,
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Navigation.java:91]: I have a sense of unease when enum constants are reordered. But it's an internal enum class, so any effects should be confined to jdk.javadoc. Thanks for making them properly ordered.

**Review by pavelrappo — APPROVED** [review]
The code in the patch looks good.

I downloaded the patch and built JDK to see the results for myself. The new page looks okay. It will take some additional effort both from jdk.javadoc and other JDK developers later to make it more useful and less silly in its content. For example, this should be the same item defined in two different places:

    Java language model  | package javax.lang.model
    Language Model       | module java.compiler

Also, I assume some doubling I see in the rightmost column will go away after #21765 has been integrated; correct?


    equivalence relation | java.lang.Object.equals(Object), java.lang.Object.equals(Object)

**Comment by hns** [comment]
> Also, I assume some doubling I see in the rightmost column will go away after #21765 has been integrated; correct?
> 
> ```
> equivalence relation | java.lang.Object.equals(Object), java.lang.Object.equals(Object)
> ```

Yes, the second entry will go away when that PR is integrated.

Thanks for reviewing!

**Pull request closed without merging**